### DISCUSSION TITLE
pygame

### DISCUSSION THREAD

**Discussion by NooRMaseR** [flet-dev/flet#3431] [Packages] [answered]
add `pygame`

**Comment by baseplate-admin** [comment]
Hi,

Sorry that i dont get the point of this.

`pygame` is sdl-2/sdl-3 and it can already be compiled for `web`/`apk` and `exe`. What's the point of `flet`?

**Reply by NooRMaseR** [reply]
When i build the app with `pygame` i get an error, that's why I add it

**Reply by kacper-the-programmer** [reply]
if I understand correctly, you wanted to build flet app with pygame in requirements.txt?

**Reply by baseplate-admin** [reply]
> When i build the app with pygame i get an error, that's why I add it

Pygame itself can be built with `python-for-android`, you dont need `flet`.

> if I understand correctly, you wanted to build flet app with pygame in requirements.txt?

Thats what i been wondering. `pygame` and `flet` are different technologies.

**Reply by NooRMaseR** [reply]
> if I understand correctly, you wanted to build flet app with pygame in requirements.txt?

Yes

**Reply by NooRMaseR** [reply]
> > When i build the app with pygame i get an error, that's why I add it
> 
> Pygame itself can be built with `python-for-android`, you dont need `flet`.
> 
> > if I understand correctly, you wanted to build flet app with pygame in requirements.txt?
> 
> Thats what i been wondering. `pygame` and `flet` are different technologies.

Yes i know, i use them together for controlling audios in the background for example text-to-speech

**Reply by kacper-the-programmer** [reply]
> Yes i know, i use them together for controlling audios in the background for example text-to-speech

Because of how `flet` and `pygame` works, it's impossible to build flet app that uses pygame

**Reply by NooRMaseR** [reply]
Ok, thanks, but do you know a better approach to control an audio file that's saved in `io.BytesIO` or a normal audio file, like pause, Continue,

**Reply by ndonkoHenri** [reply]
https://flet.dev/docs/controls/audio ?

**Reply by NooRMaseR** [reply]
sorry but this widget is not working as expected, here's an explanation of what I'm trying to do

this is the code i use when i send a request
```python
if response.get("aud"):
      self.base64_audio = base64.b64decode(response["aud"])
      pygame.mixer.music.load(BytesIO(self.base64_audio))
      pygame.mixer.music.play()
```
and works well but when i use the audio widget like this

here's the `__init__()` method

```python
self.audio = ft.Audio()
self.window.overlay.append(self.audio)
```

and here, when i get the response from the request 

```python
if response.get("aud"):
      self.audio.src_base64 = response["aud"]
      self.audio.play()
```

and i get this error `Audio must have either "src" or "src_base64" specified.` like this
![image](https://github.com/flet-dev/flet/assets/127132401/37163b6e-bc03-4a3f-b1e3-221a82d7ae74)

i tried to give it a default value like `None` or an empty `str` but same problem

**Reply by ndonkoHenri** [reply]
You could, for example, move that initialization into the `if` condition.

**Reply by NooRMaseR** [reply]
i get `Control must be added to the page first.`

here's the code
```python
if response.get("aud"):
      self.base64_audio = response["aud"]
      self.audio = ft.Audio(src_base64 = self.base64_audio, autoplay=False)
      self.window.overlay.append(self.audio)
      self.audio.play()
```

**Reply by ndonkoHenri** [reply]
Yep, you need to call a `page.update()`.
In your case it might be `self.update()` if `self` is your `Page`

**Reply by NooRMaseR** [reply]
i missed it, i have just tried it and this is the code now
```python

if response.get("aud"):
      self.base64_audio = response["aud"]
      self.audio = ft.Audio(src_base64 = self.base64_audio, autoplay=False)
      self.window.overlay.append(self.audio)
      self.window.update()
      self.audio.play()
```

and i don't get any errors, and no sound literally to hear, i even tried the `pygame` way to see if it's going to work and i get an error, here's the code:

```python
if response.get("aud"):
      self.base64_audio = response["aud"]
      self.audio = ft.Audio(src = BytesIO(b64decode(self.base64_audio)), autoplay=False)
      self.window.overlay.append(self.audio)
      self.window.update()
      self.audio.play()
```

here's the error 
```
Page error: PlatformException(WindowsAudioError, Failed to set source. For troubleshooting, see: https://github.com/bluefireteam/audioplayers/blob/main/troubleshooting.md, C:\projects\flet\client\windows\flutter\ephemeral\.plugin_symlinks\audioplayers_windows\windows\audio_player.cpp(72)\audioplayers_windows_plugin.dll!00007FFBC94A1A0D: (caller: 00007FFBC9493C8C) Exception(1) tid(3538) 8007007B The filename, directory name, or volume label syntax is incorrect.

, null)
```

**Comment by mxav1111** [comment]
I have the same issue with src_base64 where audio doesn't play and no error message/ exception generated.  Please look into this.

**Reply by ndonkoHenri** [reply]
You can open a new discussion with the details to the issue.

**Reply by mxav1111** [reply]
> You can open a new discussion with the details to the issue.

I just have a small request.  Can you please create a small working example that uses src_base64 to play audio content in buffer  I can utilize the same as my rough attempt wasn't up to the standard and then i noticed this thread.  Thanks for your help.

**Comment by ndonkoHenri** [comment, marked as answer]
To wrap up: pygame can't be built into or run inside a Flet app.

pygame is a frontend for SDL2 — it must own the native window, the render surface, and the event loop (via SDLActivity on Android and the app delegate on iOS). A Flet app is a Flutter app that already owns the screen and run-loop, so there's nowhere for SDL to attach. As @baseplate-admin noted, a pygame game is meant to be built as the app with` python-for-android`/`buildozer` — that's a standalone pygame app, not a Flet one.

For the use case in this thread (controlling audio playback) I will suggest using the [`Audio`](https://flet.dev/docs/services/audio) control: it supports `play()` / `pause()` / `resume()` / `seek()` and can play URL, files and base64.
